doc/man/man3/WinTXT.3
.TH "WinTXT" 3 "Hunter-Assassin PROJECT" \" -*- nroff -*-
.ad l
.nh
.SH NAME
WinTXT \- Une fenêtre texte est un tableau 2D de caractères\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <winTxt\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBWinTXT\fP (int dx, int dy)"
.br
.RI "Constructeur de la classe \fBWinTXT\fP\&. "
.ti -1c
.RI "void \fBclear\fP (char c=' ')"
.br
.RI "Efface le contenu de la fenêtre avec le caractère spécifié\&. "
.ti -1c
.RI "void \fBprint\fP (int x, int y, char c)"
.br
.RI "Affiche un caractère à une position spécifique dans la fenêtre\&. "
.ti -1c
.RI "void \fBprint\fP (int x, int y, char *c)"
.br
.RI "Affiche une chaîne de caractères à une position spécifique dans la fenêtre\&. "
.ti -1c
.RI "void \fBdraw\fP (int x=0, int y=0)"
.br
.RI "Dessine la fenêtre à l'écran, en commençant à la position spécifiée\&. "
.ti -1c
.RI "void \fBpause\fP ()"
.br
.RI "Met en pause le programme jusqu'à ce qu'une touche soit enfoncée\&. "
.ti -1c
.RI "char \fBgetCh\fP ()"
.br
.RI "Obtient un caractère à partir de l'entrée standard\&. "
.in -1c
.SH "Detailed Description"
.PP 
Une fenêtre texte est un tableau 2D de caractères\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "WinTXT::WinTXT (int dx, int dy)"

.PP
Constructeur de la classe \fBWinTXT\fP\&. 
.PP
\fBParameters\fP
.RS 4
\fIdx\fP Largeur de la fenêtre\&. 
.br
\fIdy\fP Hauteur de la fenêtre\&. 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "void WinTXT::clear (char c = \fR' '\fP)"

.PP
Efface le contenu de la fenêtre avec le caractère spécifié\&. 
.PP
\fBParameters\fP
.RS 4
\fIc\fP Caractère utilisé pour effacer la fenêtre (par défaut, un espace)\&. 
.RE
.PP

.SS "void WinTXT::draw (int x = \fR0\fP, int y = \fR0\fP)"

.PP
Dessine la fenêtre à l'écran, en commençant à la position spécifiée\&. 
.PP
\fBParameters\fP
.RS 4
\fIx\fP \fBPosition\fP horizontale de départ pour le dessin\&. 
.br
\fIy\fP \fBPosition\fP verticale de départ pour le dessin\&. 
.RE
.PP

.SS "char WinTXT::getCh ()"

.PP
Obtient un caractère à partir de l'entrée standard\&. 
.PP
\fBReturns\fP
.RS 4
Le caractère lu depuis l'entrée standard\&. 
.RE
.PP

.SS "void WinTXT::print (int x, int y, char * c)"

.PP
Affiche une chaîne de caractères à une position spécifique dans la fenêtre\&. 
.PP
\fBParameters\fP
.RS 4
\fIx\fP \fBPosition\fP horizontale du début de la chaîne\&. 
.br
\fIy\fP \fBPosition\fP verticale du début de la chaîne\&. 
.br
\fIc\fP Pointeur vers la chaîne de caractères à afficher\&. 
.RE
.PP

.SS "void WinTXT::print (int x, int y, char c)"

.PP
Affiche un caractère à une position spécifique dans la fenêtre\&. 
.PP
\fBParameters\fP
.RS 4
\fIx\fP \fBPosition\fP horizontale du caractère\&. 
.br
\fIy\fP \fBPosition\fP verticale du caractère\&. 
.br
\fIc\fP Caractère à afficher\&. 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for Hunter-Assassin PROJECT from the source code\&.


src/txt/winTxt.cpp
#include "winTxt.h"
#include <cassert>
#include <stdlib.h>
#include <stdarg.h>
#include <iostream>
#include <stdio.h>

#if _WIN32
#include <windef.h>
#include <winbase.h>
#include <windows.h>
#include <wincon.h>
#include <conio.h>
#else
#include <unistd.h>
#include <termios.h>
#include <unistd.h>
#endif

void termMove(int x, int y) // deplace le curseur du terminal
{
#ifdef _WIN32
    // Deplace le curseur en haut a gauche du terminal
    HANDLE console = GetStdHandle(STD_OUTPUT_HANDLE);
    COORD origine = {(SHORT)x, (SHORT)y};
    SetConsoleCursorPosition(console, origine);
#else
    char t[16];
    sprintf(t, "\033[%d;%dH", y, x);
    printf("%s", t);
#endif
}

void termClear() // efface le terminal
{
#ifdef _WIN32
    system("cls");
#else
    system("clear");
#endif
}

void termInit() // configure la saisie : ne pas afficher les caracteres tapes
{
#ifdef _WIN32
    HANDLE console = GetStdHandle(STD_INPUT_HANDLE);
    DWORD mode;
    GetConsoleMode(console, &mode);
    SetConsoleMode(console, mode & ~ENABLE_LINE_INPUT & ~ENABLE_ECHO_INPUT);
#else
    struct termios ttystate;
    bool state = true;

    // get the terminal state
    tcgetattr(STDIN_FILENO, &ttystate);

    if (state)
    {
        // turn off canonical mode
        ttystate.c_lflag &= ~ICANON;
        // minimum of number input read.
        ttystate.c_cc[VMIN] = 1;
    }
    else
    {
        // turn on canonical mode
        ttystate.c_lflag |= ICANON;
    }
    // set the terminal attributes.
    tcsetattr(STDIN_FILENO, TCSANOW, &ttystate);

    struct termios t;
    tcgetattr(STDIN_FILENO, &t);
    t.c_lflag |= ~ECHO;
    tcsetattr(STDIN_FILENO, TCSANOW, &t);
#endif
}

WinTXT::WinTXT(int dx, int dy)
{
    dimx = dx;
    dimy = dy;
    win = new char[dimx * dimy];
    clear();
    termInit();
}

void WinTXT::clear(char c)
{
    for (int i = 0; i < dimx; ++i)
        for (int j = 0; j < dimy; ++j)
            print(i, j, c);
}

void WinTXT::print(int x, int y, char c)
{
    if (x < 0)
        return;
    if (y < 0)
        return;
    if (x >= dimx)
        return;
    if (y >= dimy)
        return;
    win[y * dimx + x] = c;
}

void WinTXT::print(int x, int y, char *c)
{
    int i = 0;
    while (c[i] != '\0')
    {
        print(x + i, y, c[i]);
        ++i;
    }
}

void WinTXT::draw(int x, int y)
{
    termMove(0, 0);
    for (int j = 0; j < dimy; ++j)
    {
        for (int i = 0; i < dimx; ++i)
            printf("%c", win[j * dimx + i]);
        printf("\n");
    }
    termMove(0, dimy);
}

#if not defined _WIN32
int kbhit()
{
    struct timeval tv;
    fd_set fds;
    tv.tv_sec = 0;
    tv.tv_usec = 0;
    FD_ZERO(&fds);
    FD_SET(STDIN_FILENO, &fds); // STDIN_FILENO is 0
    select(STDIN_FILENO + 1, &fds, NULL, NULL, &tv);
    return FD_ISSET(STDIN_FILENO, &fds);
}
#endif

void WinTXT::pause()
{
#ifdef _WIN32
    system("pause");
#else
    printf("Appuyer sur une touche\n");
    while (!kbhit())
        ;
#endif
}

char WinTXT::getCh()
{ // lire un caractere si une touche a ete pressee
    char touche = 0;
#ifdef _WIN32
    if (kbhit())
    {
        DWORD mode;
        DWORD n;
        HANDLE consoleI = GetStdHandle(STD_INPUT_HANDLE);
        GetConsoleMode(consoleI, &mode);
        SetConsoleMode(consoleI, mode & ~ENABLE_LINE_INPUT & ~ENABLE_ECHO_INPUT);
        ReadConsole(consoleI, &touche, 1, &n, NULL);
    }
#else
    if (kbhit())
        touche = fgetc(stdin);
#endif
    return touche;
}
